File: src/instrument_generator.py
"""
Main Generator Orchestrator

This file coordinates between parameters, validation, and geometry generation.
You rarely need to modify this - it's the "glue" code.
"""

from typing import Dict, Any, List, Tuple
import json
from parameter_registry import get_all_output_parameters, get_derived_metadata_as_dict


def generate_violin_neck(params_json: str) -> str:
    """
    Main entry point called from JavaScript.

    Args:
        params_json: JSON string of parameter values

    Returns:
        JSON string containing:
        {
            "success": bool,
            "views": {
                "side": str,
                "top": str,
                "cross_section": str
            } | null,
            "fret_positions": {
                "available": bool,
                "html": str,
                "vsl": float,
                "no_frets": int
            } | null,
            "derived_values": dict | null,
            "errors": List[str]
        }
    """
    try:
        # Parse parameters
        params = json.loads(params_json)

        # Import here to ensure modules are loaded
        from parameter_registry import validate_parameters
        from instrument_geometry import generate_multi_view_svg, generate_fret_positions_view, calculate_derived_values

        # Validate parameters
        is_valid, errors = validate_parameters(params)

        if not is_valid:
            return json.dumps({
                "success": False,
                "views": None,
                "errors": errors
            })

        # Generate geometry (all 3 views + fret positions + derived values)
        try:
            views = generate_multi_view_svg(params)
            fret_positions = generate_fret_positions_view(params)
            derived_values = calculate_derived_values(params)

            # Format derived values for display
            output_params = get_all_output_parameters()
            formatted_values = {}
            metadata_dict = {}
            for key, value in derived_values.items():
                if key in output_params:
                    param = output_params[key]
                    decimals = param.output_config.decimals if param.output_config else 1
                    # Format value with unit
                    if param.unit:
                        formatted_values[key] = f"{value:.{decimals}f} {param.unit}"
                    else:
                        formatted_values[key] = f"{value:.{decimals}f}"
                    metadata_dict[key] = param.to_dict()

            return json.dumps({
                "success": True,
                "views": views,
                "fret_positions": fret_positions,
                "derived_values": derived_values,
                "derived_formatted": formatted_values,
                "derived_metadata": metadata_dict,
                "errors": []
            })

        except Exception as e:
            return json.dumps({
                "success": False,
                "views": None,
                "errors": [f"Geometry generation failed: {str(e)}"]
            })

    except json.JSONDecodeError as e:
        return json.dumps({
            "success": False,
            "views": None,
            "errors": [f"Invalid parameter JSON: {str(e)}"]
        })
    except Exception as e:
        return json.dumps({
            "success": False,
            "views": None,
            "errors": [f"Unexpected error: {str(e)}"]
        })


def get_derived_values(params_json: str) -> str:
    """
    Calculate derived values for the frontend with metadata.

    Args:
        params_json: JSON string of parameter values

    Returns:
        JSON string containing:
        {
            "success": bool,
            "values": {key: value},           # Raw numeric values (backward compatible)
            "metadata": {key: metadata_dict}, # Metadata for each value
            "formatted": {key: formatted_str} # Pre-formatted strings
        }
    """
    try:
        from instrument_geometry import calculate_derived_values

        params = json.loads(params_json)
        derived_raw = calculate_derived_values(params)

        # Build enhanced response with metadata
        output_params = get_all_output_parameters()
        formatted_values = {}
        metadata_dict = {}

        for key, value in derived_raw.items():
            if key in output_params:
                param = output_params[key]
                decimals = param.output_config.decimals if param.output_config else 1
                # Format value with unit
                if param.unit:
                    formatted_values[key] = f"{value:.{decimals}f} {param.unit}"
                else:
                    formatted_values[key] = f"{value:.{decimals}f}"
                metadata_dict[key] = param.to_dict()

        return json.dumps({
            "success": True,
            "values": derived_raw,           # Backward compatible: raw numbers
            "metadata": metadata_dict,        # NEW: metadata for each value
            "formatted": formatted_values     # NEW: pre-formatted strings
        })
    except Exception as e:
        return json.dumps({
            "success": False,
            "errors": [str(e)]
        })


def get_derived_value_metadata() -> str:
    """
    Get metadata definitions for all derived values.

    Returns:
        JSON string containing:
        {
            "success": bool,
            "metadata": {key: metadata_dict}
        }
    """
    try:
        return json.dumps({
            "success": True,
            "metadata": get_derived_metadata_as_dict()
        })
    except Exception as e:
        return json.dumps({
            "success": False,
            "errors": [str(e)]
        })


def get_parameter_definitions() -> str:
    """
    Get parameter definitions for UI generation.

    Returns:
        JSON string of parameter definitions
    """
    try:
        from parameter_registry import get_parameters_as_json
        return get_parameters_as_json()
    except Exception as e:
        return json.dumps({
            "success": False,
            "errors": [f"Failed to load parameters: {str(e)}"]
        })


def get_ui_metadata() -> str:
    """
    Get complete UI metadata bundle including sections, presets, parameters, and derived values.

    This is the single source of truth for UI organization.

    Returns:
        JSON string containing:
        {
            "success": bool,
            "metadata": {
                "sections": {...},
                "presets": {...},
                "parameters": {...},
                "derived_values": {...}
            }
        }
    """
    try:
        from ui_metadata import get_ui_metadata_bundle
        return json.dumps({
            "success": True,
            "metadata": get_ui_metadata_bundle()
        })
    except Exception as e:
        return json.dumps({
            "success": False,
            "errors": [str(e)]
        })


def get_presets() -> str:
    """
    Get available parameter presets from the presets directory.

    Returns:
        JSON string of presets dictionary: { preset_id: { name: str, parameters: dict } }
    """
    try:
        import os
        presets = {}

        # Get path to presets directory (relative to this file)
        current_dir = os.path.dirname(os.path.abspath(__file__))
        presets_dir = os.path.join(current_dir, '..', 'presets')

        # Check if presets directory exists
        if not os.path.exists(presets_dir):
            return json.dumps({})

        # Load all JSON files from presets directory
        for filename in os.listdir(presets_dir):
            if filename.endswith('.json'):
                filepath = os.path.join(presets_dir, filename)
                try:
                    with open(filepath, 'r') as f:
                        preset_data = json.load(f)

                    # Extract parameters and instrument name
                    if 'parameters' in preset_data:
                        params = preset_data['parameters']
                        instrument_name = params.get('instrument_name', filename.replace('.json', ''))

                        # Use filename without extension as preset ID
                        preset_id = filename.replace('.json', '')
                        presets[preset_id] = {
                            'name': instrument_name,
                            'parameters': params
                        }
                except Exception as e:
                    print(f"Warning: Failed to load preset {filename}: {str(e)}")
                    continue

        return json.dumps(presets)
    except Exception as e:
        return json.dumps({
            "success": False,
            "errors": [f"Failed to load presets: {str(e)}"]
        })


if __name__ == '__main__':
    # Test the generator
    from parameter_registry import get_default_values
    import json
    
    print("Testing violin neck generator...")
    print("=" * 50)
    
    # Get defaults
    defaults = get_default_values()
    params_json = json.dumps(defaults)
    
    print("\n1. Testing parameter definitions...")
    param_defs = get_parameter_definitions()
    print(f"✓ Loaded {len(json.loads(param_defs)['parameters'])} parameters")
    
    print("\n2. Testing generation...")
    result = generate_violin_neck(params_json)
    result_data = json.loads(result)

    if result_data['success']:
        print(f"✓ Generation successful!")
        views = result_data['views']
        print(f"  Side view: {len(views['side'])} bytes")
        print(f"  Top view: {len(views['top'])} bytes")
        print(f"  Cross-section: {len(views['cross_section'])} bytes")
    else:
        print(f"✗ Generation failed:")
        for error in result_data['errors']:
            print(f"  - {error}")
    
    print("\n3. Testing presets...")
    presets = get_presets()
    preset_data = json.loads(presets)
    print(f"✓ Loaded {len(preset_data)} presets")
    
    print("\n" + "=" * 50)
    print("All tests complete!")

File: src-ts/instrument_generator.ts
/**
 * Overstand - Instrument Generator (TypeScript port)
 *
 * Main generator orchestrator - coordinates between parameters, validation,
 * and geometry generation.
 *
 * Ported from src/instrument_generator.py
 *
 * NOTE ON API SURFACE:
 *
 * The JSON-based functions (generateViolinNeck, getDerivedValues, etc.) accept
 * and return JSON strings to match the Python API surface exactly - this keeps
 * JavaScript call sites unchanged when switching from Pyodide to TypeScript.
 *
 * The object-based generateViolin function is provided for direct TypeScript
 * use (e.g. parity tests) and accepts/returns plain objects.
 */

import {
  validateParameters,
  getParametersAsJson,
  getDerivedMetadataAsDict,
  getAllOutputParameters,
} from './parameter_registry'

import {
  calculateDerivedValues,
  generateMultiViewSvg,
} from './instrument_geometry'

import { getUiMetadataBundle } from './ui_metadata'

import { calculateFretPositions } from './geometry_engine'

export { loadStencilFont } from './radius_template'

import type { Params } from './geometry_engine'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface GenerateViolinResult {
  success: boolean
  views: Record<string, string> | null
  fret_positions: FretPositionsResult | null
  derived_values: Record<string, number | null> | null
  derived_formatted: Record<string, string> | null
  derived_metadata: Record<string, unknown> | null
  errors: string[]
}

export interface FretPositionsResult {
  available: boolean
  html?: string
  vsl?: number
  no_frets?: number
  message?: string
}

export interface DerivedValuesResult {
  success: boolean
  values?: Record<string, number | null>
  formatted?: Record<string, string>
  metadata?: Record<string, unknown>
  errors?: string[]
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * Generate fret positions view (HTML table + metadata).
 * Mirrors view_generator.py generate_fret_positions_view().
 */
function generateFretPositionsView(params: Params): FretPositionsResult {
  const vsl = (params['vsl'] as number) || 0
  const instrumentFamily = (params['instrument_family'] as string) || 'VIOLIN'
  let noFrets: number

  if (params['no_frets'] != null) {
    noFrets = params['no_frets'] as number
  } else if (instrumentFamily === 'VIOL') {
    noFrets = 7
  } else if (instrumentFamily === 'GUITAR_MANDOLIN') {
    noFrets = 20
  } else {
    noFrets = 0
  }

  if (noFrets === 0) {
    return { available: false, message: 'Fret positions not applicable for violin family' }
  }

  const fretPositions = calculateFretPositions(vsl, noFrets)

  let html = '<div class="fret-table-container">'
  html += '<table class="fret-table">'
  html +=
    '<thead><tr><th>Fret</th><th>Distance from Nut (mm)</th><th>Distance from Previous Fret (mm)</th></tr></thead>'
  html += '<tbody>'

  let prevPos = 0
  for (let i = 0; i < fretPositions.length; i++) {
    const pos = fretPositions[i]!
    const fretNum = i + 1
    const fromNut = pos
    const fromPrev = pos - prevPos
    html += `<tr><td>${fretNum}</td><td>${fromNut.toFixed(1)}</td><td>${fromPrev.toFixed(1)}</td></tr>`
    prevPos = pos
  }

  html += '</tbody></table></div>'

  return {
    available: true,
    html,
    vsl,
    no_frets: noFrets,
  }
}

/**
 * Build formatted derived values and metadata dict.
 */
function buildFormattedAndMetadata(
  derivedValues: Record<string, number | null>
): {
  formattedValues: Record<string, string>
  metadataDict: Record<string, unknown>
} {
  const outputParams = getAllOutputParameters()
  const formattedValues: Record<string, string> = {}
  const metadataDict: Record<string, unknown> = {}

  for (const [key, value] of Object.entries(derivedValues)) {
    if (value === null) continue
    const param = outputParams[key]
    if (!param || !param.output_config) continue

    const decimals = param.output_config.decimals
    const formatted = value.toFixed(decimals)
    if (param.unit) {
      formattedValues[key] = `${formatted} ${param.unit}`
    } else {
      formattedValues[key] = formatted
    }
    metadataDict[key] = param.toOutputMetadata()
  }

  return { formattedValues, metadataDict }
}

// ---------------------------------------------------------------------------
// generateViolin (object-based, for TypeScript direct use / parity tests)
// ---------------------------------------------------------------------------

/**
 * Generate all views and derived values for an instrument.
 *
 * Accepts and returns plain objects (not JSON strings).
 * This is the main TypeScript entry point, e.g. for parity tests.
 */
export function generateViolin(params: Params): GenerateViolinResult {
  try {
    const [isValid, errors] = validateParameters(params)

    if (!isValid) {
      return {
        success: false,
        views: null,
        fret_positions: null,
        derived_values: null,
        derived_formatted: null,
        derived_metadata: null,
        errors,
      }
    }

    try {
      const views = generateMultiViewSvg(params)
      const fretPositions = generateFretPositionsView(params)
      const derivedValues = calculateDerivedValues(params)

      const { formattedValues, metadataDict } = buildFormattedAndMetadata(derivedValues)

      return {
        success: true,
        views,
        fret_positions: fretPositions,
        derived_values: derivedValues,
        derived_formatted: formattedValues,
        derived_metadata: metadataDict,
        errors: [],
      }
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e)
      const isGeometryError =
        msg.includes('math domain error') ||
        msg.includes('Invalid geometric constraints')
      const errorMsg = isGeometryError
        ? 'Invalid geometric constraints - check that neck dimensions, angles, and measurements are compatible. Common issues: bridge height too low, neck angle too steep, or fingerboard radius too small for the width.'
        : `Geometry generation failed: ${msg}`

      return {
        success: false,
        views: null,
        fret_positions: null,
        derived_values: null,
        derived_formatted: null,
        derived_metadata: null,
        errors: [errorMsg],
      }
    }
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    return {
      success: false,
      views: null,
      fret_positions: null,
      derived_values: null,
      derived_formatted: null,
      derived_metadata: null,
      errors: [`Unexpected error: ${msg}`],
    }
  }
}

// ---------------------------------------------------------------------------
// JSON-based API (mirrors Python instrument_generator.py exactly)
// ---------------------------------------------------------------------------

/**
 * Main generation entry point - called by generation.js generateNeck().
 *
 * Accepts and returns JSON strings, matching the Python generate_violin_neck() API.
 * Strips _generator_url before processing (it's added by generation.js but is
 * not a real instrument parameter).
 */
export function generateViolinNeck(paramsJson: string): string {
  try {
    const params = JSON.parse(paramsJson) as Params

    // Strip internal metadata keys not part of the instrument parameter set
    if ('_generator_url' in params) {
      delete (params as Record<string, unknown>)['_generator_url']
    }

    const result = generateViolin(params)

    return JSON.stringify(result)
  } catch (e) {
    if (e instanceof SyntaxError) {
      return JSON.stringify({
        success: false,
        views: null,
        errors: [`Invalid parameter JSON: ${e.message}`],
      })
    }
    const msg = e instanceof Error ? e.message : String(e)
    return JSON.stringify({
      success: false,
      views: null,
      errors: [`Unexpected error: ${msg}`],
    })
  }
}

/**
 * Calculate derived values only (no SVG rendering).
 *
 * Returns JSON: { success, values, formatted, metadata }
 * Mirrors Python get_derived_values().
 */
export function getDerivedValues(paramsJson: string): string {
  try {
    const params = JSON.parse(paramsJson) as Params
    const derivedRaw = calculateDerivedValues(params)

    const { formattedValues, metadataDict } = buildFormattedAndMetadata(derivedRaw)

    return JSON.stringify({
      success: true,
      values: derivedRaw,
      metadata: metadataDict,
      formatted: formattedValues,
    })
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    return JSON.stringify({
      success: false,
      errors: [msg],
    })
  }
}

/**
 * Get metadata definitions for all derived/output values.
 *
 * Returns JSON: { success, metadata }
 * Mirrors Python get_derived_value_metadata().
 */
export function getDerivedValueMetadata(): string {
  try {
    return JSON.stringify({
      success: true,
      metadata: getDerivedMetadataAsDict(),
    })
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    return JSON.stringify({
      success: false,
      errors: [msg],
    })
  }
}

/**
 * Get parameter definitions for UI generation.
 *
 * Returns JSON of parameter definitions.
 * Mirrors Python get_parameter_definitions().
 */
export function getParameterDefinitions(): string {
  try {
    return getParametersAsJson()
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    return JSON.stringify({
      success: false,
      errors: [`Failed to load parameters: ${msg}`],
    })
  }
}

/**
 * Get complete UI metadata bundle.
 *
 * Returns JSON: { success, metadata: { sections, presets, parameters, derived_values } }
 * Mirrors Python get_ui_metadata().
 */
export function getUiMetadata(): string {
  try {
    return JSON.stringify({
      success: true,
      metadata: getUiMetadataBundle(),
    })
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e)
    return JSON.stringify({
      success: false,
      errors: [msg],
    })
  }
}
